### PR TITLE
raftstore: check range before replica a peer

### PR DESCRIPTION
**Pull request by BusyJay**
ref #1217 

@siddontang @hhkbp2 PTAL

### CODE REVIEW

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -791,6 +786,17 @@ impl Peer {
         send_msg.set_from_peer(from_peer);
         send_msg.set_to_peer(to_peer);
 
+        if self.get_store().is_initialized() &&
+           (msg_type == MessageType::MsgRequestVote ||
+            // the peer has not been known to this leader, it may exist or not.
+            msg_type == MessageType::MsgHeartbeat && msg.get_commit() == INVALID_INDEX) {
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: better to use `(msg_type == MessageType::MsgHeartbeat && msg.get_commit() == INVALID_INDEX)`

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -500,12 +501,37 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+        let msg_type = message.get_msg_type();
+        if msg_type != MessageType::MsgRequestVote &&
+           (msg_type != MessageType::MsgHeartbeat || message.get_commit() != INVALID_INDEX) {
+            warn!("target peer {:?} doesn't exist, stale message {:?}.",
+                  target,
+                  msg_type);
+            return Ok(false);
+        }
+
+        let start_key = data_key(msg.get_start_key());
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: No, if they are both empty, it will not pass the following check.

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -470,7 +470,8 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         self.register_raft_base_tick(event_loop);
     }
 
-    fn check_target_peer_valid(&mut self, region_id: u64, target: &metapb::Peer) -> Result<bool> {
+    fn check_target_peer_valid(&mut self, region_id: u64, msg: &RaftMessage) -> Result<bool> {
```
Comment (ngaut) [inline src/raftstore/store/store.rs]: s/check_target_peer_valid/maybe_create_peer/

File: src/raftstore/store/store.rs (hunk 2 above)
Comment (ngaut) [inline src/raftstore/store/store.rs]: It would be more clean if we check both start key and end key.

Comment (BusyJay) [inline src/raftstore/store/store.rs]: `end_key` is checked on L523.

Comment (siddontang) [inline src/raftstore/store/store.rs]: If the start and end key are both empty, we must receive an invalid message, not a overlapped message.

File: src/raft/mod.rs [hunk 4]
```diff
@@ -37,7 +37,7 @@ mod read_only;
 
 pub use self::storage::{RaftState, Storage};
 pub use self::errors::{Result, Error, StorageError};
-pub use self::raft::{Raft, StateRole, Config, INVALID_ID, SoftState};
+pub use self::raft::{Raft, StateRole, Config, INVALID_ID, INVALID_INDEX, SoftState};
```
Comment (BusyJay) [inline src/raft/mod.rs]: It doesn't only stand for commit_index.

Comment (ngaut) [inline src/raft/mod.rs]: Then we should define two constants to make it more clear.

Comment (BusyJay) [inline src/raft/mod.rs]: What if we want to use INVALID_INDEX for first_index and last_index? I think `INVALID_INDEX` is ok, commit_index is one type of index, INVALID_INDEX is also an invalid commit index.

File: src/raftstore/store/store.rs (hunk 2 above)
Comment (ngaut) [inline src/raftstore/store/store.rs]: That's wired, I suggest we set the start key of first region to 0x0, and set the end key to 0xFF.
I just reviewed the currently code, our tests rely on empty start key and end key.

**Comment by siddontang** [comment]
LGTM

File: src/raftstore/store/store.rs (hunk 2 above)
Comment (BusyJay) [inline src/raftstore/store/store.rs]: Which region are supposed to cover the key `0xFFFF`? I think this has been discussed long time ago. @siddontang

**Comment by ngaut** [comment]
PTAL @hhkbp2

**Comment by BusyJay** [comment]
That's not true. The first region's start_key and end_key are both empty.

File: src/raftstore/store/store.rs [hunk 5]
```diff
@@ -470,7 +470,12 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         self.register_raft_base_tick(event_loop);
     }
 
-    fn check_target_peer_valid(&mut self, region_id: u64, target: &metapb::Peer) -> Result<bool> {
+    /// If target peer doesn't exist, create one.
```
Comment (hhkbp2) [inline src/raftstore/store/store.rs]: s/create one/create it/

File: src/raftstore/store/store.rs [hunk 6]
```diff
@@ -528,7 +558,7 @@ impl<T: Transport, C: PdClient> Store<T, C> {
             return Ok(());
         }
 
-        if !try!(self.check_target_peer_valid(region_id, msg.get_to_peer())) {
+        if !try!(self.maybe_create_peer(region_id, &msg)) {
```
Comment (hhkbp2) [inline src/raftstore/store/store.rs]: Considering the context and its function, IMO the original name is a better name.

File: src/raftstore/store/peer.rs [hunk 7]
```diff
@@ -791,6 +786,17 @@ impl Peer {
         send_msg.set_from_peer(from_peer);
         send_msg.set_to_peer(to_peer);
 
+        if self.get_store().is_initialized() &&
+           (msg_type == MessageType::MsgRequestVote ||
+            // the peer has not been known to this leader, it may exist or not.
```
Comment (hhkbp2) [inline src/raftstore/store/peer.rs]: This comment is not quite useful. Comments occur in this code block should give a plain explanation on the purpose of setting region start key and end key in these messages. Something like:

```If a leader doesn't have the commit index info for a peer, there could be two cases:
1. That peer already exists but has not establish communication with leader yet
2. That peer is added newly due to member change or region split, but it's not created yet
For case 2, the region start key and end key are attached in RequestVote and Heartbeat message for the  store of that peer to check whether to create a new peer for member change when receiving these messages, or just to wait for a pending region split to perform later.
```

**Comment by hhkbp2** [comment]
LGTM

**Pull request merged**